website/content/en/docs/about/under-the-hood/architecture/buffering-model.md

```diff
@@ -0,0 +1,215 @@
+running Vector, crashed. While pull-based sources like S3 or Kafka would handle this by simply
+reattempting to process the events, push-based sources may not be able to retransmit their messages.
+
+### Disk buffers
+
+When the durability of data is more important than the overall performance of Vector, disk buffers
+can be used to persist buffered events between stopping and starting Vector, including when
+Vector crashes. Disk buffers allow Vector to essentially pick up from where it left off when it
+starts back up again.
+
+Disk buffers function like a write-ahead log, where every event is first sent through the buffer,
+and written to the data files, before it is read back out. This may sound slow, but in practice,
+modern operating systems allow reads to happen out of memory, so disk buffers generally maintain
+high throughput on both the read and write path. By default, we do not synchronize data to disk for
+every write, but instead synchronize on an interval (500 milliseconds) which allows for high
+throughput with a reduced risk of data loss.
+
+We've designed disk buffers to provide _consistent_ performance. While other projects may be able to
+write data to disk faster than Vector, we've chosen to make sure that events can be read as fast as
+they can be written, as well as reducing the tail latencies between an event being written and an
+event being read on the other side.
+
+Additionally, like in-memory buffers, disk buffers have a configurable maximum size so they can be
+limited in terms of disk usage. This maximum size is adhered to rigidly, so you can depend on Vector
+not exceeding it. There is a minimum size for all buffers, though -- currently ~256MiB -- which is a
+requirement of the disk buffer implementation. On the filesystem, disk buffers will look like
+append-only log files that grow to a maximum file size of 128MiB and are deleted once they have been
+processed fully.
+
+Storage errors are always a potential issue, whether due to hardware failures or data files being
+mistakenly deleted while Vector is running. Disk buffers automatically checksum all events being
+written to disk, and when corruption is detected during a read, they will automatically recover as
+many events as can be correctly decoded. Disk buffers will also emit metrics when such corruption is
+detected, to give as accurate of a view into the number of events that were lost as it possibly can.
+
+#### Operator requirements
+
+{{< warning >}}
+Disk buffers have some unique monitoring requirements compared to in-memory buffers, specifically
+around free storage space.
+{{< /warning >}}
+
+In order to provide the durability guarantees that an event written to a disk buffer is safely on
+disk, many common error handling techniques cannot be used, such as retrying failed operations and
+so on. This means that Vector will **forcefully panic** and exit the process when an I/O error
```

Comment:
So, essentially: there's no way to recover when we encounter an I/O error during flush. We have to panic, and restart Vector, which lets us reopen the buffer and scan through, figuring out if we need to start writing fresh to a new data file, etc.

The recent work I did was simply to avoid the ugly panic messages we would previously emit, and turn those into proper error messages that bubbled up the topology, etc.

Reply:
👍 it would generally be good to be clear between panic and error, I've seen more than a few bug reports claiming Vector panicked, when it simply errored and exited.

I feel like panic has more negative "out-of-control" connotations.

Reply:
Yeah, fair point. I can change this to "exit".